### ISSUE TITLE
Weighted loss function

### ISSUE THREAD

**Issue by pow3rpi** [ultralytics/ultralytics#20022]
### Search before asking

- [x] I have searched the Ultralytics [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

A year or 2 ago I saw a discussion (here in issues) about how to apply/implement weighted loss function in YOLO models to deal with class imbalance (I experienced the same problem). As I can see today, it hasn't been resolved still, so people have to manually edit source code to use it in their projects (you have to do it after each library installation). If you still cannot configure weights via data.yml or model parameters I would be glad to do it for the community and to implement this feature.

### Use case

Class imbalance

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @pow3rpi, thank you for your interest in Ultralytics 🚀 and for your willingness to contribute to the project!

Weighted loss functions to address class imbalance are indeed an important topic, and we appreciate your initiative in offering to implement this feature for the community. To ensure we fully understand the scope of your proposal, please provide any additional details or specific ideas you have for the feature implementation. For example, how you envision configuring the class weights (e.g., via `data.yml`, model parameters, or another method).

For reference, you may want to explore our [Docs](https://docs.ultralytics.com/) for information on our library's current structure and functionalities. If you're ready to proceed, you can submit your feature through a Pull Request (PR). To get started, please review our [Contributing Guidelines](https://github.com/ultralytics/ultralytics/blob/main/CONTRIBUTING.md) for best practices.

In the meantime, we recommend the following actions to ensure your environment is up to date:

## Upgrade

Upgrade to the latest `ultralytics` package along with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of the following verified environments with dependencies preinstalled:

- **Notebooks** with free GPU:  
  <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM: See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI: See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**: See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/)  
  <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Community Support

For further discussions and assistance, join our Ultralytics community:
- [Discord](https://discord.com/invite/ultralytics) for real-time chat 🎧
- [Discourse](https://community.ultralytics.com/) for in-depth discussions
- [Subreddit](https://reddit.com/r/Ultralytics) to share knowledge with others

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests ensure the reliability of YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) across macOS, Windows, and Ubuntu.

This is an automated response, but don't worry—an Ultralytics engineer will review your issue and provide further guidance soon. Thank you for contributing and being part of the Ultralytics community! 🚀✨

**Comment by glenn-jocher** [comment]
@pow3rpi thank you for your interest in enhancing YOLO's capabilities! Our current loss implementations like FocalLoss already support class weighting through the `alpha` parameter, which can help address class imbalance. You can customize this by modifying your model's loss configuration or subclassing the loss function as shown in the Advanced Customization guide. 

We welcome PRs that implement this in a maintainable way without requiring source edits post-installation. Please ensure any changes maintain backward compatibility and follow our Contributing Guide. The community will appreciate your contribution! 🚀

**Comment by pow3rpi** [comment]
@glenn-jocher thenk you for clarification. I will take a look.

**Comment by glenn-jocher** [comment]
Sounds good, @pow3rpi. Feel free to ask if any specific questions come up during your review.

**Comment by pow3rpi** [comment]
@glenn-jocher do u have a working example how to apply FocalLoss in current YOLOv11 implementation? Struggling to find it in the docs. I assume that it still should be done manually...

**Comment by glenn-jocher** [comment]
@pow3rpi The FocalLoss is already implemented in the YOLO codebase (as seen in the `utils/loss.py` file), but you're right that accessing it requires some customization. To apply it in YOLO11, you'll need to create a custom trainer that overrides the default loss function:

```python
from ultralytics import YOLO
from ultralytics.models.yolo.detect import DetectionTrainer
from ultralytics.utils.loss import FocalLoss

class CustomDetectionTrainer(DetectionTrainer):
    def get_model(self, cfg, weights):
        model = super().get_model(cfg, weights)
        # Customize model's loss function
        model.model[-1].loss_fn = FocalLoss(gamma=2.0, alpha=0.25)  # Adjust parameters as needed
        return model

# Train with custom trainer
model = YOLO("yolo11n.pt")
results = model.train(trainer=CustomDetectionTrainer, data="your_dataset.yaml", epochs=50)
```

This approach allows you to inject the FocalLoss implementation without permanently modifying the source code. The alpha parameter is especially useful for class imbalance as it provides weights for positive/negative examples.

**Comment by glenn-jocher** [comment]
You're absolutely right, @pow3rpi. Looking at the `FocalLoss` implementation in our codebase, `gamma` and `alpha` are defined in the `forward` method, not in `__init__`. This means the parameters can't be set during instantiation as I suggested.

For now, you have two options:
1. Create a custom loss class that inherits from `FocalLoss` with your desired parameters
2. Set the parameters when the loss is called in the training loop

Moving these parameters to the `__init__` method would be a good improvement to make the API more intuitive and consistent with other PyTorch modules. If you'd like to submit a PR that modifies `FocalLoss` to accept these parameters during initialization while maintaining backward compatibility, that would be a valuable contribution to the codebase.

**Comment by pow3rpi** [comment]
@glenn-jocher Yeah, I will do it. Thank you for your confirmation.

**Comment by glenn-jocher** [comment]
Great to hear, @pow3rpi! That would be a valuable improvement to make the API more intuitive and consistent. When working on the PR, consider maintaining backward compatibility by making gamma and alpha optional parameters in `__init__` with the same default values (gamma=1.5, alpha=0.25) as currently defined in the forward method. If you need any guidance during the implementation or want to discuss your approach before submitting, feel free to reach out. Looking forward to your contribution!

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in [PR #20292](https://github.com/ultralytics/ultralytics/pull/20292), titled **`ultralytics 8.3.116` RTDETR VarifocalLoss `gamma` and `alpha` parameterization**. Huge thanks to @pow3rpi for authoring the PR, with valuable contributions from @glenn-jocher and @Laughing-q! 🎉

### Key Changes in the PR:
- Introduced `gamma` and `alpha` parameters to **FocalLoss** and **VarifocalLoss**, enabling more control over class imbalance and focusing on harder predictions.
- Enhanced **DETRLoss** to support these parameters, improving flexibility and fine-tuning potential.
- Made loss functions more user-friendly by allowing instance-level parameterization instead of static settings.
- Updated the version to `8.3.116`.

These updates enhance model training flexibility and performance, particularly for imbalanced datasets or challenging classification problems. The improvements may directly address the issues reported here.

### How to Test:
To test if the changes resolve your issue, you can update your package to the latest version including this fix:
```bash
pip install -U ultralytics>=8.3.116
```
Alternatively, you can install the latest code directly from the main branch:
```bash
pip install git+https://github.com/ultralytics/ultralytics.git@main
```
You can also wait for the next official release on PyPI if preferred.

Let us know if the updates resolve your issue! If you encounter any further problems, feel free to share feedback here. Thank you for reporting this and helping improve the project! 🙏

**Comment by glenn-jocher** [comment]
I can confirm that after the PR #20292 was merged, both `FocalLoss` and `VarifocalLoss` now properly accept `gamma` and `alpha` parameters in their `__init__` methods, as seen in the reference documentation. Now you can use them directly without subclassing:

```python
from ultralytics import YOLO
from ultralytics.models.yolo.detect import DetectionTrainer
from ultralytics.utils.loss import FocalLoss

class CustomDetectionTrainer(DetectionTrainer):
    def get_model(self, cfg, weights):
        model = super().get_model(cfg, weights)
        # Now we can properly pass parameters during initialization
        model.model[-1].loss_fn = FocalLoss(gamma=2.0, alpha=0.25)
        return model
```

Thanks for highlighting this issue and submitting the fix. This makes the API much more intuitive and consistent with PyTorch's design patterns.

**Comment by Y-T-G** [comment]
Duplicate of https://github.com/ultralytics/ultralytics/issues/21082

**Comment by Jordan-Pierce** [comment]
@Y-T-G @glenn-jocher  

I'm working on implementing a loss function for dealing with [partially labeled data](https://link.springer.com/article/10.1007/s11263-023-01755-4), and I'm finding that no matter what changes I make, they do not seem to stick even though training is working fine, and classes (via print statements) are being correctly initialized. 

Is there something in the training pipeline that overwrites / doesn't respect changes somehow? What am I missing? More annoying is that the results are identical to a baseline I'm training (in another notebook). 

It's fine that the loss isn't identical to the original research, I'm just confused as to why they are always the same?

```python

import torch
import torch.nn as nn
from ultralytics.models.yolo.detect.train import DetectionTrainer
from ultralytics.models.yolo.detect.loss import v8DetectionLoss

# ==============================================================================
# CLASS 1: NegativeRegionDroppingLoss 
# ==============================================================================
class NegativeRegionDroppingLoss(nn.Module):
    def __init__(self, drop_rate: float = 0.5):
        super().__init__()
        self.base_loss_fn = nn.BCEWithLogitsLoss(reduction='none')
        self.drop_rate = drop_rate

    def forward(self, pred: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
        loss = self.base_loss_fn(pred, target)
        
        if not self.training or self.drop_rate <= 0:
            return loss.mean()

        negative_mask = (target == 0)
        if negative_mask.sum() == 0:
            return loss.mean()

        rand_mask = torch.rand_like(loss, device=loss.device) < self.drop_rate
        drop_mask = negative_mask & rand_mask
        keep_mask = ~drop_mask
        
        loss = loss[keep_mask]
        
        if loss.numel() == 0:
            return torch.tensor(0.0, device=pred.device, requires_grad=True)

        return loss.mean()

# ==============================================================================
# CLASS 2: The Custom Detection Loss
# ==============================================================================
class CustomDetectionLoss(v8DetectionLoss):
    def __init__(self, model):
        super().__init__(model)
        drop_rate = getattr(model.args, 'nrd_drop_rate', 0.5)
        self.nrd_loss = NegativeRegionDroppingLoss(drop_rate=drop_rate)
        print(f"CustomDetectionLoss initialized with NRD drop_rate: {drop_rate}")

    def __call__(self, preds, batch):
        # Manually set training mode for our submodule
        self.nrd_loss.train(self.model.training)

        loss = torch.zeros(3, device=self.device)
        feats = preds[1] if isinstance(preds, tuple) else preds
        pred_distri, pred_scores = torch.cat([xi.view(feats[0].shape[0], self.no, -1) for xi in feats], 2).split(
            (self.reg_max * 4, self.nc), 1)

        pred_scores = pred_scores.permute(0, 2, 1).contiguous()
        pred_distri = pred_distri.permute(0, 2, 1).contiguous()

        dtype = pred_scores.dtype
        batch_size = pred_scores.shape[0]
        imgsz = torch.tensor(feats[0].shape[2:], device=self.device, dtype=dtype) * self.stride[0]
        anchor_points, stride_tensor = make_anchors(feats, self.stride, 0.5)

        targets = torch.cat((batch["batch_idx"].view(-1, 1), batch["cls"].view(-1, 1), batch["bboxes"]), 1)
        targets = self.preprocess(targets.to(self.device), batch_size, scale_tensor=imgsz[[1, 0, 1, 0]])
        gt_labels, gt_bboxes = targets.split((1, 4), 2)
        mask_gt = gt_bboxes.sum(2, keepdim=True).gt_(0)

        pred_bboxes = self.bbox_decode(anchor_points, pred_distri)

        _, target_bboxes, target_scores, fg_mask, _ = self.assigner(
            pred_scores.detach().sigmoid(), (pred_bboxes.detach() * stride_tensor).type(gt_bboxes.dtype),
            anchor_points * stride_tensor, gt_labels, gt_bboxes, mask_gt)

        target_scores_sum = max(target_scores.sum(), 1)

        # Instead of calling the parent's loss, we implement our own here.

        # 1. Classification Loss (The part we want to change)
        # Use our custom NRD loss function directly on the scores.
        loss[1] = self.nrd_loss(pred_scores, target_scores.to(dtype))

        # 2. Bbox and DFL Loss (Keep the original logic for these)
        if fg_mask.sum():
            target_bboxes /= stride_tensor
            loss[0], loss[2] = self.bbox_loss(
                pred_distri, pred_bboxes, anchor_points, target_bboxes, target_scores, target_scores_sum, fg_mask)

        # Apply the hyperparameter gains
        loss[0] *= self.hyp.box
        loss[1] *= self.hyp.cls
        loss[2] *= self.hyp.dfl

        # Return final loss
        return loss.sum() * batch_size, loss.detach()

# ==============================================================================
# CLASS 3: The Custom Trainer
# ==============================================================================
class CustomTrainer(DetectionTrainer):
    def get_model(self, cfg=None, weights=None, verbose=True):
        model = super().get_model(cfg, weights, verbose)
        if not hasattr(model, 'args'):
            model.args = self.args
        
        # We need to replace the loss criterion on the Detect() head.
        detect_head = model.model[-1]
        detect_head.loss = CustomDetectionLoss(model)
        
        return model


args = dict(
    model="yolov8n.pt",
    data=dataset,
    project=project,
    name="PU_Loss",
    task='detect',
    epochs=10,
    half=True,
    imgsz=640,
    single_cls=True,
    plots=True,
    batch=16,
    workers=0 
)

trainer = CustomTrainer(overrides=args)
trainer.train()

```

**Comment by Y-T-G** [comment]
Should be patching the `criterion` attribute.

`model.criterion = CustomDetectionLoss(model)`
